suite/tests/linux/threadset.c

```diff
@@ -0,0 +1,345 @@
+#    else
+#        define MOVE_TO_YMM(buf, num)                           \
+            __asm__ __volatile__("vmovdqu %0, %%ymm" #num       \
+                                 :                              \
+                                 : "m"(buf[num * INTS_PER_YMM]) \
+                                 : "ymm" #num);
+        MOVE_TO_YMM(buf, 0)
+        MOVE_TO_YMM(buf, 1)
+        MOVE_TO_YMM(buf, 2)
+        MOVE_TO_YMM(buf, 3)
+        MOVE_TO_YMM(buf, 4)
+        MOVE_TO_YMM(buf, 5)
+        MOVE_TO_YMM(buf, 6)
+        MOVE_TO_YMM(buf, 7)
+#        ifdef X64
+        MOVE_TO_YMM(buf, 8)
+        MOVE_TO_YMM(buf, 9)
+        MOVE_TO_YMM(buf, 10)
+        MOVE_TO_YMM(buf, 11)
+        MOVE_TO_YMM(buf, 12)
+        MOVE_TO_YMM(buf, 13)
+        MOVE_TO_YMM(buf, 14)
+        MOVE_TO_YMM(buf, 15)
+#        endif
+#    endif
+
+        write(2, "before\n", 7);
+
+        /* Sometime in this loop, we will synch with the other thread */
+        for (int i = 0; i < 100; i++) {
+            dummy2();
+        }
+
+        write(2, "after\n", 6);
+
+        /* Ensure they are preserved across the sigreturn (xref i#3812). */
+#    ifdef __AVX512F__
+        /* Use a new buffer to avoid the old values. We could do a custom memset
+         * with rep movs in asm instead (regular memset may clobber SIMD regs).
+         */
+        int buf2[INTS_PER_ZMM * NUM_SIMD_AVX512_REGS];
+#        define MOVE_FROM_ZMM(buf, num)                          \
+            __asm__ __volatile__("vmovdqu64 %%zmm" #num ", %0"   \
+                                 : "=m"(buf[num * INTS_PER_ZMM]) \
+                                 :                               \
+                                 : "zmm" #num);
+        MOVE_FROM_ZMM(buf2, 0)
+        MOVE_FROM_ZMM(buf2, 1)
+        MOVE_FROM_ZMM(buf2, 2)
+        MOVE_FROM_ZMM(buf2, 3)
+        MOVE_FROM_ZMM(buf2, 4)
+        MOVE_FROM_ZMM(buf2, 5)
+        MOVE_FROM_ZMM(buf2, 6)
+        MOVE_FROM_ZMM(buf2, 7)
+#        ifdef X64
+        MOVE_FROM_ZMM(buf2, 8)
+        MOVE_FROM_ZMM(buf2, 9)
+        MOVE_FROM_ZMM(buf2, 10)
+        MOVE_FROM_ZMM(buf2, 11)
+        MOVE_FROM_ZMM(buf2, 12)
+        MOVE_FROM_ZMM(buf2, 13)
+        MOVE_FROM_ZMM(buf2, 14)
+        MOVE_FROM_ZMM(buf2, 15)
+        MOVE_FROM_ZMM(buf2, 16)
+        MOVE_FROM_ZMM(buf2, 17)
+        MOVE_FROM_ZMM(buf2, 18)
+        MOVE_FROM_ZMM(buf2, 19)
+        MOVE_FROM_ZMM(buf2, 20)
+        MOVE_FROM_ZMM(buf2, 21)
+        MOVE_FROM_ZMM(buf2, 22)
+        MOVE_FROM_ZMM(buf2, 23)
+        MOVE_FROM_ZMM(buf2, 24)
+        MOVE_FROM_ZMM(buf2, 25)
+        MOVE_FROM_ZMM(buf2, 26)
+        MOVE_FROM_ZMM(buf2, 27)
+        MOVE_FROM_ZMM(buf2, 28)
+        MOVE_FROM_ZMM(buf2, 29)
+        MOVE_FROM_ZMM(buf2, 30)
+        MOVE_FROM_ZMM(buf2, 31)
+#        endif
+        for (i = 0; i < NUM_SIMD_AVX512_REGS; i++) {
+            for (j = 0; j < INTS_PER_ZMM; j++) {
+                if (buf2[i * INTS_PER_ZMM + j] != 0xdeadbeef + i * INTS_PER_ZMM + j) {
```

Comment:
nit: Why not `assert(buf2[i * INTS_PER_ZMM + j] == 0xdeadbeef + i * INTS_PER_ZMM + j)`?
Same below (line 333).

Reply:
before the patch is applied, it segfaults in `assert` before printing anything

Reply:
But as above: we have the patch, so we're not running this test w/o the patch.  So we can use assert() and print(), right?